### PR TITLE
Versions using layouts and bootstrap

### PR DESCRIPTION
**Pull request by dgrammatiko**
### Use of layouts for field contenthistory

This is a redone for #4561. Now versions are using layouts and there are two sets:
One on the root/layouts this is for B/C
And another on the templates isis and protostar this is using bootstrap modal
### Actual rendering:

Isis:
screen shot 2015-01-09 at 10 07 13

Protostar:
screen shot 2015-01-09 at 10 08 01
### B/C

None
### Testing

Try to re edit an article on backend and front end (make sure that versions are enabled and you have few versions available)

### CODE REVIEW

**Comment by wilsonge** [comment]
Note to committers DO NOT merge this for 3.4. Wait until me and @phproberto have worked out the final desired form for a rendering class.

**Comment by dgt41** [comment]
@wilsonge I am not trying to change your (and @phproberto ’s) timelines here!
My idea here was that if introducing: user, media and versions fields using layouts and bootstrap and also have #4661, #4645, #4563 and #5652 merged for 3.4 then the team could advertise that Joomla dropped almost totally Mootools in core.
BUT that could happen ONLY IF the fields DO NOT introduce a wrong approach which then will become a pain in the @$$ to try to maintain, update etc. 
Also I don’t know what are the USP s for 3.4, but I assume that advertising a total transition to jQuery could be a nice feature (amongst the others)

Bottom line: if you want me to put it on hold (e.g. close it for now) I will so

**Comment by wilsonge** [comment]
I completely understand the reasons for this! And I support killing off mootools as well. The issue for me is how we intialise and render the JLayout. The framework View package from 2.0 will likely include this renderer class Michael wrote for Jissues (see ). We had a JLayout implementation here https://github.com/joomla-projects/cms-naked/blob/master/libraries/cms/renderer/jlayout.php and were thinking about using that. 

But my point is that whilst we still are really unsure about how any rendering system will work in Joomla I'm really reluctant to start doing everything before we have a 'accepted' solution. Because otherwise we start getting inconsistencies and b/c breaks. I'll try and get @phproberto to comment on this before you actually close it. But my opinion is to (reluctantly) hold it .

**Comment by phproberto** [comment]
Thanks @wilsonge for taking care of this. I've moved all my github notifications to another folder and it's harder for me to notice things like this.

Yes. I think this should be moved to the v3.5 release. I've just had a phone call with Javier to prepare the v3.5 enviroment. I think we are going to have the same setup than v3.4 so we will have a milestone + a tag for PRs related to it. Once we have settled where & how is v3.5 code going to be managed @wilsonge and I can have the definitive renderer ready and then start telling people how integrate things like this PR. 

I don't think that this PR will need a lot of changes, just use the new rendering system. I agree that it makes no sense to introduce something now to then change it on v3.5. 

So I think we should keep this PR open until in around 2 weeks we have things settled. Then point it to the right place with the right instructions.

Thanks @dgt41 !!

**Comment by dgt41** [comment]
@wilsonge @phproberto Excellent! FYI there is also a media field layout, but I forgot to push the code here, so you don’t have to redo that one as well 😄. As soon as you settle the 3.5-dev repo I will bring it there!

One more think that needs your attention is #5652 which is a redo of the current implementation with a little twist to have bootstrap modal in the equation. And the question, as well there, is the same shall it wait for 3.5, or can it go in 3.4?

**Comment by Fedik** [comment]
**test** works good

**Comment by anibalsanchez** [comment]
#Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by roland-d** [comment]
@wilsonge Have you figured things out as to what to do with this PR? Can this be merged into 3.5 or should it wait longer?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by wilsonge** [comment]
If we implement the stuff Roberto mentioned in https://github.com/joomla/joomla-cms/pull/7948#issuecomment-144778913 then I'm happy (this also applies for #5655)

**Comment by dgt41** [comment]
To test this try to edit an article in isis, hathor, protostar and beez

File: libraries/cms/form/field/contenthistory.php [hunk 1]
```diff
@@ -25,31 +25,45 @@ class JFormFieldContenthistory extends JFormField
 	public $type = 'ContentHistory';
 
 	/**
-	 * Method to get the content history field input markup.
+	 * Layout to render the label
 	 *
-	 * @return  string  The field input markup.
+	 * @var  string
+	 */
+	protected $renderLayout = 'joomla.form.field.contenthistory';
```
Comment (zero-24) [inline libraries/cms/form/field/contenthistory.php]: is there a reason to use `$renderLayout` and not `$layout` as e.g. here: https://github.com/joomla/joomla-cms/pull/7702/files#diff-f4e4dfd1f5119bfe9d55b7c74570c7a6R42 ?

**Comment by zero-24** [comment]
Expected my comment here. Hathor is broken:

hathor_broken

and in beez we have no option to use the versions anyway :smile: (bevor and after the patch)

protostar and isis looks good to me.

File: libraries/cms/form/field/contenthistory.php (hunk 1 above)
Comment (dgt41) [inline libraries/cms/form/field/contenthistory.php]: good point

**Comment by dgt41** [comment]
@zero-24 Fixed!

**Comment by zero-24** [comment]
@dgt41 i don't see any change (on github)? Are you sure you commit the change for $renderLayout and $layout. Did you also catch that hathor problem?

**Comment by dgt41** [comment]
@zero-24 yes both hathor and the $layout are now fixed

**Comment by zero-24** [comment]
Great i will try it tomorrow. Thanks.

**Comment by zero-24** [comment]
I have tested this item :red_circle: unsuccessfully on f45ad43791d9b05c5238aedc1e715add786f1325<br /><br />hmm i have still the same problems with hathor here :( Sorry for the delay @dgt41 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by designbengel** [comment]
I have tested this item :red_circle: unsuccessfully on f45ad43791d9b05c5238aedc1e715add786f1325<br /><br />I have also issues in Hathor - it´s cut off, when the window is resized. And despite i don´t think anyone would use this on mobile - it seems to be broken also. (screens following)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by designbengel** [comment]
bildschirmfoto 2015-11-01 um 17 39 46
bildschirmfoto 2015-11-01 um 17 38 06

**Comment by dgt41** [comment]
@designbengel Hathor and Beez are not really compatible with bootstrap. Let them use mootools…
 So this PR now converts the content history modal to bootstrap ONLY for Isis and Protostar

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 9c5bef64debc3e8e57bb662d173e3b5674efa1ef<br /><br />Works now or lets better say it is broken the same way as bevor for hathor. :smiley: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on c1596fb9224d206f3445747139f02c997e9aa4e5<br /><br />Still works with hathor and isis as well as protostar. for testing this https://github.com/joomla/joomla-cms/pull/8248 needs to be applyed as well.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by phproberto** [comment]
I have tested this item :white_check_mark: successfully on bc4331dcf7478b0f1c22fb79f0a4212a7ac4341f<br /><br />Works perfectly. Requires the modal CSS fixes from https://github.com/joomla/joomla-cms/pull/5871 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by designbengel** [comment]
At squeezing the window there is some stuff cut off.
bildschirmfoto 2015-11-05 um 19 29 52

**Comment by waader** [comment]
I have tested this item :white_check_mark: successfully on bc4331dcf7478b0f1c22fb79f0a4212a7ac4341f<br /><br />As @designbengel noted: there are cut off´s when reducing the viewport, but on the other hand they are also present in the mootools version. So being pragmatic I would suggest to go with the current version and maybe improve here and there later.

I tested with IE8 and current FF, Chrome and IE but with no mobile phone.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by dgt41** [comment]
Added a span with class hidden phone to make the buttons fit the mobile view. Preview:
screenshot 2015-11-05 20 37 31

@waader @designbengel can you confirm that is fixed now?

**Comment by waader** [comment]
Great! Before retesting ... the checkbox on left side of each row is also "hidden". Maybe you have another idea to get that fixed?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by dgt41** [comment]
@waader I didn’t realize that in order to compare two versions you need to be able to select them. Now should be ok

**Comment by designbengel** [comment]
I have tested this item :white_check_mark: successfully on ba74815dc9a92d58a75eb7e255168732ecbeb41b<br /><br />now everything is nice. 
Suggestion: There is a close button at the bottom. In the banner section it´s a cancel button. maybe same wording for all modals?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5654">issues.joomla.org/joomla-cms/5654</a>.</sub>

**Comment by waader** [comment]
Now it's definitely better than the mootools version. Thanks!

**Comment by wilsonge** [comment]
Merged - two successful tests from @waader and @designbengel
